Sources/Instrumentation/Tracing/Span.swift

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Tracing open source project
+//
+// Copyright (c) 2020 Moritz Lang and the Swift Tracing project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Baggage
+import Dispatch
+
+public protocol Span {
+    var operationName: String { get }
+
+    var startTimestamp: DispatchTime { get }
+    var endTimestamp: DispatchTime? { get }
+
+    var baggage: BaggageContext { get }
+
+    var events: [SpanEvent] { get }
+    mutating func addEvent(_ event: SpanEvent)
+
+    var attributes: [String: SpanAttribute] { get }
+    subscript(attributeName attributeName: String) -> SpanAttribute? { get set }
+
+    var onEnd: (Span) -> Void { get }
+    mutating func end(at timestamp: DispatchTime)
+}
+
+extension Span {
+    public func addingEvent(_ event: SpanEvent) -> Self {
+        var copy = self
+        copy.addEvent(event)
+        return copy
+    }
+
+    public mutating func end() {
+        self.end(at: .now())
+    }
+}
+
+public struct SpanEvent {
+    public let name: String
+    public let timestamp: DispatchTime
+
+    public init(name: String, at timestamp: DispatchTime = .now()) {
+        self.name = name
+        self.timestamp = timestamp
+    }
+}
+
+public enum SpanAttribute {
+    case string(String)
+    case int(Int)
+    case double(Double)
+    case bool(Bool)
+    case array([SpanAttribute])
+    case any(CustomStringConvertible)
+}
+
```

Comment:
Should we also offer:

```swift
// MARK: Span Attributes: Semantic Conventions

// https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md
extension SpanAttribute {
  // Transport protocol used. See note below.
  public func netTransport(_ value: String) -> SpanAttribute {
    .string(value)
  }
```

Reply:
And offer all those https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/README.md

Reply:
I was already thinking about whether we should bake all of these semantics into the `Instrument` library or make them available separately (domain-specific) instead. What's your opinion about this?

Reply:
I see, you mean since that's OpenTelemetry, that's a good point. I wonder, technically we are pretty much conforming anyway to OT here, so maybe can do this as well... 

Can do as a separate ticket, could you ticketify?

Reply:
https://github.com/slashmo/gsoc-swift-tracing/issues/65